[Ma Yuehan] iP

MaYuehan opened the pull request:
# DukeProject 

DukeProject is a task management application designed to simplify your life. It offers a range of features to help you stay organized and productive.

>“Your mind is for having ideas, not holding them.” – David Allen

**DukePro frees your mind of having to remember things you need to do. It's,**

- Text-based interface for quick task management
- Easy to learn and use
- Super fast performance



**All you need to do is:**

1. Download it from [here](https://github.com/MaYuehan/ip/releases/tag/A-JUnit).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉 

And it is __FREE__!

## Key Features
- [x] Task tracking and management
- [x] Deadline management


If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Duke {

    private Storage storage;
    private TaskList tasks;
    private Ui ui;

    /**
     * Constructs a Duke object with the specified file path.
     *
     * @param filePath The path to the file where tasks are stored.
     */
    public Duke(String filePath) {
        ui = new Ui();
        storage = new Storage(filePath);
        try {
            tasks = new TaskList(storage.load());
        } catch (DukeException e) {
            ui.showLoadingError();
            tasks = new TaskList(new ArrayList<>());
        }
    }

    /**
     * Runs the chatbot, displaying welcome message and processing user commands.
     */
    public void run() {
        ui.showWelcome();
        boolean isExit = false;
        while (!isExit) {
            try {
                String fullCommand = ui.readCommand();
                Command c = Parser.parse(fullCommand);
                c.execute(tasks, ui, storage);
                isExit = c.isExit();
            } catch (DukeException e) {
                ui.showError(e.getMessage());
            }
        }
    }

    public static void main(String[] args) {
        new Duke("./data/botYue.txt").run();
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Commit:
Implement an initial skeletal version of it that simply greets the user and exits.Change name to BotYue

Commit:
Improve the skeletal version of Duke so that it echos commands entered by the user, and exits when the user types the command bye.

Commit:
Add the ability to store whatever text entered by the user and display them back to the user when requested.

Commit:
Add the ability to mark tasks as done. Optionally, add the ability to change the status back to not done.

Commit:
Add support for tracking three types of tasks: ToDos: tasks without any date/time attached to it
Deadlines: tasks that need to be done before a specific date/time
Events: tasks that start at a specific date/time and ends at a specific date/time

Commit:
fix the identation, add the input and expected output for A-TextUiTesting

Commit:
Teach the chatbot to deal with errors such as incorrect inputs entered by the user.

Commit:
update the A-TextUiTesting

Commit:
Add support for deleting tasks from the list.

Commit:
update the A-TextUiTesting for level 6

Commit:
Use Java enum

Commit:
update enum

Commit:
update Text UI Testing for enum

Commit:
refactor

Commit:
create new classes and file for level7

Commit:
Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when the chatbot starts up.

Commit:
level 7 update

Commit:
update level7

Commit:
update

Commit:
Teach the chatbot how to understand dates and times.

Commit:
update

Commit:
Refactor the code to extract out closely related code as classes.

Commit:
add packages

Commit:
add package2

Commit:
update

Commit:
Merge commit 'a6f73244a20539e20c2189973ebf597df34ef631'

Commit:
A-Gradle

Commit:
Add JUnit tests

Commit:
Add JavaDoc comments

Commit:
add javadoc comment

Commit:
Tweak the code to comply with a coding standard

Commit:
conding standard update

Commit:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Command/MarkCommand.java

Commit:
Give users a way to find a task by searching for a keyword.

Commit:
Update the description of the ip so that it contains the following GFMD elements:a heading, a bullet list, a numbered list, a fenced code block (with syntax highlighting), a task list, an emoji, a blockquote, a hyperlink, inline code, some text formatting: bold, italic, strikethrough etc.

Commit:
update the readme file of this project, add the main classes into the readme file.

Commit:
Add a GUI to the chatbot. Use the JavaFX technology to implement the GUI.

Commit:
Add a GUI to the chatbot. Use the JavaFX technology to implement the GUI.

Commit:
import the images into the resources folder.

ChenXy128 commented on src/main/java/duke/Command/Command.java:
```diff
@@ -0,0 +1,30 @@
+/**
+ * Represents a command to be executed by Duke.
+ */
+public abstract class Command {
```
Nice abstract class!🤩

Commit:
update GUI: delete the Ui class, add the show the message during the command.

ChenXy128 commented:
Excellent work! Just gotta abide by the coding standard a little more and it will be perfect😇

moguries commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,208 @@
+        String command = parts[0].toLowerCase();
+
+        switch (command) {
+            case "bye":
```
I think this format doesn't comply with the style guide! (no indentations for 'case' clauses)

moguries commented on src/main/java/duke/Tasks/TaskList.java:
```diff
@@ -0,0 +1,73 @@
+     *
+     * @return The number of tasks.
+     */
+    public int size() {
```
Perhaps this method could be named `getSize` to follow the verb naming way ?

moguries reviewed (commented):
LGTM :)

Commit:
update the exception part, implement the exception function into the GUI (when input an invalid command, show the exception on the user interface)

Commit:
delete redundant classes and parts of code that is no longer needed any more to make the code be simpler and clearer.

Commit:
solve the corresponding errors that caused by deleting the Ui class.

Commit:
Use assert feature (not JUnit assertions) to document important assumptions that should hold at various points in the code.

Commit:
Critically examines the code and refactor to improve the code quality where necessary. Avoid magic numbers, avoid complicated expressions, especially those having many negations and nested parentheses.

Commit:
Lay out the code so that it adheres to the logical structure. Adjust the code to achieve higher code quality.

MaYuehan committed:
Merge branch 'master' into A-CodeQuality

MaYuehan committed:
Merge pull request #1 from MaYuehan/A-CodeQuality

Lay out the code so that it adheres to the logical structure. Adjust …

Commit:
Merge branch 'master' of github.com:MaYuehan/ip

# Conflicts:
#	src/main/java/duke/Command/MarkCommand.java

Commit:
solve the merge conflict by delete the redundant code piece

Commit:
Refactor Storage class: Enhance save() function

1. Revamped the save() function in the Storage class to improve code
readability, error handling, and resource management.

2. Implemented try-with-resources for FileWriter to ensure proper closing of
the resource.
3. Streamlined the logic for checking task types during saving by removing
redundant condition checks.
4. Utilized String.format() for constructing task details to enhance
readability and maintainability.
5. Simplified the method signature by removing unnecessary parameters.

Commit:
Detect and Remove Duplicate Tasks

The new `DetectCommand` class is introduced to address the need for detecting and removing duplicate tasks from the task list. This command iterates through the task list, comparing each task's description character by character with descriptions of other tasks. If a duplicate is found, it adds the task to a list of duplicate tasks. After detecting duplicates, the command removes them from the task list and saves the updated task list to storage.

Changes made:
- Added a new `DetectCommand` class to the project.
- Implemented logic to compare task descriptions character by character and remove duplicates from the task list.
- Improved the `execute` method to provide feedback on duplicate removal, including listing duplicate tasks if any.
- Introduced a helper method `compareDescriptions` to facilitate character-by-character comparison of task descriptions.

The introduction of the `DetectCommand` class enhances the functionality of the application by ensuring task lists remain free of duplicates, leading to improved task management and user experience.

Commit:
Improve the GUI to make it more polished. Added more specific exception message information to inform the user that what is the exact problem of their command input and provide them the correct command formate so they can quickly fix the input into right format. Also crop the image of the duke and user to a circle to make it has a better UI and more like a chat interface.
In addition, some of the issues that shown in the github issue boxes has been fixed. for example, the boolean value name marked is fixed to isMarked to follow the coding standard.

Commit:
Create the BotYue user guide in the Readme file. Include the overall key features, the instruction of adding different types of tasks, the instruction of greeting and exiting the chatbot. Provide a very specific and clear command instruction for the users of using all the functionalities of this project. Also, the error handling is also included in the user guide to clarify the situation when exceptions thrown.

Commit:
Give a unique personality. Deleted the original DialogBox class and fxml, created two new classes, UserDialogBox and YueDialogBox, to modified the user interface separately. The dialog box is made as in a chating panel of instagram or wechat, also the font and the colour of the text, the background of the dialog container is also changed.
beside modified the GUI, I also add a new method called parseFindCommand to handle the error for find command, and trim the command to avoid throwing exception with extra whitespace.

Commit:
rename the class Duke to class Yue, and rename the duke folder and all the corresponding use. And the Ui.pnd image into the user guide to show the GUI window.
Added a time scheduler if the user input the bye command, the system will end the run in 1 second.

Commit:
Update the user guide of the outside readme, the one not in the docs folder.  Include the overall key features, the instruction of adding different types of tasks, the instruction of greeting and exiting the chatbot. Provide a very specific and clear command instruction for the users of using all the functionalities of this project. Also, the error handling is also included in the user guide to clarify the situation when exceptions thrown.

Commit:
Add the line enableAssertions = true in the the run section of the build.gradle to make sure the jar file can open and run after it generate. Also,  the Yue class is been modified to initialise the parser input.

Commit:
as the line enableAssertions = true added in the build.gradle causes a lot of assertion error, it is then removed from the file and unchange the corresponding change made because of this line.

Commit:
Firstly, a Junit test is been added to the test folder to examine the functionality of class EventTask, especially the parse time and formateDateTime function of both the inputed start time and end time. Then, I rename the Command package and the Tasks package to the lowercases, command and tasks. This is because the name of the packages is suggested to be in lowercase as following the package naming convention specified by the coding standard.

Commit:
Add JUnit tests for TodoCommand and Parser.
This commit adds comprehensive JUnit tests for the `TodoCommand` class and the `Parser` class. Implemented JUnit tests for the `execute` method of the `TodoCommand` class to ensure that it correctly adds a new todo task to the task list. Added test cases to cover different scenarios such as valid input, null input, blank input, and unknown commands in the `ParserTest` class. Ensured that the tests provide adequate coverage for the parsing logic and exception handling in the `Parser` class. All tests are passing, confirming the correct behavior of the `TodoCommand` and `Parser` classes under various conditions.
These tests enhance the reliability and maintainability of the codebase by ensuring that critical functionality related to task creation and command parsing functions as expected. They also serve as documentation for future developers, providing clarity on expected behavior and edge cases handled by these components.
Also, the error message of input index that is out of range is changed to: the task index is out of range, of mark, unmark and delete command. The corresponding modification in the user guide is also made to align with the change.

Commit:
Fix the configuration error
Modify the build.gradle file, update the dependencies by reusing Woojiahao's code snippet to fix the warning, unsupported JavaFX configuration: classes were loaded from 'unnamed module @282f3878'.

Commit:
Modify the load() method in Storage class
As the shadowJar does not pack the data directory with the main folder, the load() method is updated to create the directory and the file to be written when the system cannot detect the data folder and file exit.

Commit:
Update the readme file
Add quick start instruction for the user to guide them how to install and run the program before get into it.  Provide the link of the jar file that should be downloaded to run the program. Also, bolden the command format text part to make them be more obvious, make the user guide be more user-friendly.

Commit:
Refine the code quality
Remove the chunks of commented out code. Rename the Duke to Yue in every javaDoc comment. Reduce the length of some very long method. As some int parameter has name of uppercase letters, the name is then changed into lowercase letters. Update some comments of the method that is modified.Add the comment for the code that is reused from other students.

Commit:
Divide very-long method into two methods
The load() method is too long, it is divided into two shorter methods to improve the code quality. Also, the execute method in the DetectCommand class is also divided into two shorter methods to make the code be neat and improve the code quality.

Commit:
Update the test folder
As some classes such as Parser are modified in the main folder, the corresponding classes in the test folder is updated as well. Regenerate the jar file by using command ShadowJar.

Commit:
Update the Readme
The screenshot of the Ui.png is in the folder of docs, which means it cannot accessed by the outside Readme file. Hence, update the code to docs/Ui.png to help it to access the Ui.png.

MaYuehan committed:
Update README.md 

Update the Find tasks part to make the description be more detailed as the find command is case sensitive and not full-text search.

MaYuehan committed:
Update README.md

Add more details about the find command as it is case sensitive and not full-text earch.
